extensions/iceberg/src/main/java/io/deephaven/iceberg/util/IcebergCatalogAdapter.java

```diff
@@ -830,4 +884,559 @@ private Table readTableInternal(
+        }
+        final Schema schema = new Schema(fields);
+
+        final PartitionSpec partitionSpec = createPartitionSpec(schema, partitioningColumnNames);
+        return new SpecAndSchema(partitionSpec, schema);
+    }
+
+    private static PartitionSpec createPartitionSpec(
+            @NotNull final Schema schema,
+            @NotNull final Iterable<String> partitionColumnNames) {
+        final PartitionSpec.Builder partitionSpecBuilder = PartitionSpec.builderFor(schema);
+        for (final String partitioningColumnName : partitionColumnNames) {
+            partitionSpecBuilder.identity(partitioningColumnName);
+        }
+        return partitionSpecBuilder.build();
+    }
+
+    private org.apache.iceberg.Table createNewIcebergTable(
+            @NotNull final TableIdentifier tableIdentifier,
+            @NotNull final SpecAndSchema specAndSchema,
+            @NotNull final IcebergParquetWriteInstructions writeInstructions) {
+        if (catalog.tableExists(tableIdentifier)) {
+            throw new IllegalArgumentException("Table already exists: " + tableIdentifier);
+        }
+        return catalog.createTable(tableIdentifier, specAndSchema.schema, specAndSchema.partitionSpec, null,
+                Map.of(TableProperties.DEFAULT_FILE_FORMAT, DEFAULT_GENERATED_FILE_FORMAT,
+                        TableProperties.PARQUET_COMPRESSION,
+                        StringUtils.toLowerCase(writeInstructions.compressionCodecName())));
+    }
+
+    private static class CompletedParquetWrite {
+        private final URI destination;
+        private final long numRows;
+        private final long numBytes;
+
+        private CompletedParquetWrite(final URI destination, final long numRows, final long numBytes) {
+            this.destination = destination;
+            this.numRows = numRows;
+            this.numBytes = numBytes;
+        }
+    }
+
+    @NotNull
+    private static List<CompletedParquetWrite> writeParquet(
+            @NotNull final org.apache.iceberg.Table icebergTable,
+            @NotNull final Table[] dhTables,
+            @NotNull final IcebergParquetWriteInstructions writeInstructions) {
+        // Build the parquet instructions
+        final List<CompletedParquetWrite> parquetFilesWritten = new ArrayList<>(dhTables.length);
+        final ParquetInstructions.OnWriteCompleted onWriteCompleted =
+                (destination, numRows, numBytes) -> parquetFilesWritten
+                        .add(new CompletedParquetWrite(destination, numRows, numBytes));
+        final ParquetInstructions parquetInstructions = writeInstructions.toParquetInstructions(
+                onWriteCompleted, icebergTable.schema().idToName());
+
+        // Write the data to parquet files
+        for (final Table dhTable : dhTables) {
+            final long epochMicrosNow = Instant.now().getEpochSecond() * 1_000_000 + Instant.now().getNano() / 1_000;
+            final String filename = new StringBuilder()
```

Comment:
Does DH care about the filesystem order? I would have expected that we always read wrt some order from iceberg manifests?

Stylistically, it might be nice to always write to increasing filenames, but does our correctness _depend_ on it?

Reply:
For reference, `org.apache.iceberg.io.OutputFileFactory#generateFilename`:

```java
  private String generateFilename() {
    return format.addExtension(
        String.format(
            "%05d-%d-%s-%05d%s",
            partitionId,
            taskId,
            operationId,
            fileCount.incrementAndGet(),
            null != suffix ? "-" + suffix : ""));
  }
```

Reply:
Our correctness does not depend on the order but a fixed order makes it easier to test.

Some new things I learned:
All the discovered parquet files are read in parallel (your change only in SourceTable::maybeAddLocations) and then they are ordered to make a corresponding rowset. Check RegionColumnSourceManager around line which does `entriesToInclude.stream().sorted().forEachOrdered`.
For this ordering, we use `URITableLocationKey::compareTo` method, which either checks the order varaible (if set, we were not setting it) or checks the URI, etc.

So now, I have set the order for each discovered key so that the order in which data files is discovered is the order in which we add them to the table.
And i have changed the naming logic to not use the time but just a counter. Therefore, all files added in a single write call will be ordered, but we cannot comment across write calls.